### PR TITLE
[WIP] Initial alternative plugin system

### PR DESCRIPTION
**Pull request by KrauseFx**
Alternative to https://github.com/fastlane/fastlane/pull/4619

### CODE REVIEW

**Comment by KrauseFx** [comment]
I haven't yet looked into how we can actually load and use those gems then. 
I was assuming it would be nice to have separate lock files, but thinking about it, that's not what lock files are built-for.

I'll iterate once again, and use only one lock file. We can still have separate Gemfiles, so that it's easy for us to provide a CLI to add and remove plugins. 
All we have to do is add this to the `./Gemfile`:

```ruby
content = File.read(File.join(File.dirname(__FILE__), 'fastlane', 'FastlaneGemfile'))
eval(content, binding)
```

this way we still have separate Gemfiles, one in the `fastlane` directory, but only have one lock file.

This technique however doesn't allow us to update just the plugins without actually updating all the other things, right?

File: fastlane/lib/fastlane/commands_generator.rb [hunk 1]
```diff
@@ -181,8 +183,16 @@ def run
 
         c.action do |args, options|
           plugin_name = args.last # TODO: actually parse the things
+          plugin_name = 'fastlane_' + plugin_name unless plugin_name.start_with?('fastlane_')
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: Maybe this should actually be part of the `PluginManager`

File: fastlane/lib/fastlane/plugin_manager.rb [hunk 2]
```diff
@@ -2,35 +2,43 @@ module Fastlane
...
-      nil
+      File.read(pluginsfile_path) if File.exist?(pluginsfile_path)
     end
 
     def add_dependency(plugin_name)
       unless (pluginsfile_content || "").include?(plugin_name)
         content = pluginsfile_content || "# Autogenerated by fastlane\n\n"
-        content += "gem '#{plugin_name}'\n"
+        # TODO: obviously need to stop hardcoding a git URL!
+        content += "gem '#{plugin_name}', git: 'git@github.com:fastlane/#{plugin_name}.git'\n"
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/plugin_manager.rb]: By default, we'll just have `gem "fastlane_name"`, right? The git url should only be used when the owner doesn't push their gem to RubyGems

Comment (KrauseFx) [inline fastlane/lib/fastlane/plugin_manager.rb]: And during development this will be a local path

File: fastlane/lib/fastlane/plugin_manager.rb [hunk 3]
```diff
@@ -40,56 +48,65 @@ def add_dependency(plugin_name)
...
+        UI.important("Please add the following code to '#{path_to_gemfile}':")
         puts ""
         puts code_to_attach.magenta # we use puts to make it easier to copy and paste
-        UI.user_error!("Please update your Gemfile and run fastlane again")
+        UI.user_error!("Please update '#{path_to_gemfile} and run fastlane again")
       end
       return true
     end
 
+    def create_gemfile(path)
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/plugin_manager.rb]: Maybe add a `!` operator, to be consistent with `attach_plugins!`, both methods change local files

File: fastlane/lib/fastlane/plugin_manager.rb (hunk 2 above)
Comment (mfurtak) [inline fastlane/lib/fastlane/plugin_manager.rb]: Yes, as the comment says, the git business should not be there in the long term!

File: fastlane/lib/fastlane/commands_generator.rb (hunk 1 above)
Comment (mfurtak) [inline fastlane/lib/fastlane/commands_generator.rb]: Yup, seems right

File: fastlane/bin/fastlane [hunk 4]
```diff
@@ -1,5 +1,13 @@
 #!/usr/bin/env ruby
 $LOAD_PATH.push File.expand_path("../../lib", __FILE__)
 
+require 'fastlane_core'
```
Comment (KrauseFx) [inline fastlane/bin/fastlane]: We have to make sure to not forget the rocket emoji 🚀

Comment (KrauseFx) [inline fastlane/bin/fastlane]: Oh, you thought of it, never mind 👍

**Comment by jcampbell05** [comment]
Excited for this! Are you able to give an overview how this will work so I have some context when reviewing :)

File: fastlane/lib/fastlane/plugin_manager.rb [hunk 5]
```diff
@@ -24,8 +25,12 @@ def pluginsfile_content
       File.read(pluginsfile_path) if pluginsfile_path && File.exist?(pluginsfile_path)
     end
 
+    def self.plugin_prefix
+      FASTLANE_PLUGIN_PREFIX
+    end
+
     def add_dependency(plugin_name)
-      plugin_name = 'fastlane_' + plugin_name unless plugin_name.start_with?('fastlane_')
+      plugin_name = self.class.plugin_prefix + plugin_name unless plugin_name.start_with?(plugin_prefix)
```
Comment (mfurtak) [inline fastlane/lib/fastlane/plugin_manager.rb]: `plugin_prefix` is being called two different ways on this line

**Comment by KrauseFx** [comment]
Closing this PR to split up the work 👍

**Pull request closed without merging**